File: backend/compact_generation.py
import re
import json
def find_unique_alphabet_letters(premises):
    unique_letters = set()

    for premise in premises:
        letters = re.findall(r'[A-Za-z]+', premise)
        for letter in letters:
            unique_letters.add(letter)

    return unique_letters


def find_x_non_unique_letters(premises, x):
    unique_letters = find_unique_alphabet_letters(premises)
    non_unique_letters = []
    letter_ord = ord('A')

    while len(non_unique_letters) < x:
        if chr(letter_ord) not in unique_letters:
            non_unique_letters.append(chr(letter_ord))
        letter_ord += 1

    return non_unique_letters

def toggle_negation(expression):
    if expression.startswith("¬"):
        return expression[1:] if len(expression) == 2 else expression[2:-1]
    else:
        return "¬" + expression if len(expression) == 1 else "¬(" + expression + ")"
    
def modus_tollens(premise_to_replace, var2):
    var1 = toggle_negation(premise_to_replace)
    return {
        "premise1": toggle_negation(var2),
        "premise2": f"{var1}→{var2}",
        "conclusion": premise_to_replace
    }


def modus_ponens(premise_to_replace, var2):
    return {
        "premise1": var2,
        "premise2": f"{var2}→{premise_to_replace}",
        "conclusion": premise_to_replace
    }


def disjunctive_syllogism(premise_to_replace, var2):
    return {
        "premise1": f"{var2}∨{premise_to_replace}",
        "premise2": toggle_negation(var2),
        "conclusion": premise_to_replace
    }

def use_given_rule(premises, premise_to_replace, rule_func, justification):
    non_unique_letters = find_x_non_unique_letters(premises, 1)
    var2 = non_unique_letters[0]
    filled_rule = rule_func(premise_to_replace, var2)
    index_to_replace = next(i for i, step in enumerate(solution) if step['statement'] == premise_to_replace)

    del solution[index_to_replace]
    solution.insert(index_to_replace, {
        "stepID": index_to_replace,
        "statement": filled_rule['premise1'],
        "justification": "Premise",
        "evidence": ""
    })
    solution.insert(index_to_replace + 1, {
        "stepID": index_to_replace + 1,
        "statement": filled_rule['premise2'],
        "justification": "Premise",
        "evidence": ""
    })
    solution.insert(index_to_replace + 2, {
        "stepID": index_to_replace + 2,
        "statement": premise_to_replace,
        "justification": justification,
        "evidence": f"{index_to_replace + 1}, {index_to_replace + 2}"
    })

    for i in range(index_to_replace + 3, len(solution)):
        solution[i]['stepID'] = i
        if solution[i]['evidence']:
            numbers = list(map(int, solution[i]['evidence'].split(',')))
            updated_numbers = [(number + 2 if number >= index_to_replace + 1 else number) for number in numbers]
            solution[i]['evidence'] = ','.join(map(str, updated_numbers))

    premises.remove(premise_to_replace)
    premises.append(filled_rule['premise1'])
    premises.append(filled_rule['premise2'])
    return premises

def use_modus_tollens(premises, premise_to_replace, solution):
    return use_given_rule(premises, premise_to_replace, modus_tollens, "Modus Tollens")


def use_modus_ponens(premises, premise_to_replace):
    return use_given_rule(premises, premise_to_replace, modus_ponens, "Modus Ponens")


def use_disjunctive_syllogism(premises, premise_to_replace):
    return use_given_rule(premises, premise_to_replace, disjunctive_syllogism, "Disjunctive Syllogism")

def pretty_print_solution(solutions):
    print("-------------------------------------------------")
    for step in solutions:
        print(f"Step {step['stepID']}:", step['statement'])
        print("  Justification:", step['justification'])
        if step['evidence']:
            print("  Evidence:", step['evidence'])
    print("-------------------------------------------------")

# Example usage
premises = ["P", "P→Q"]
conclusion = "Q"
solution = [
    {"stepID": 0, "statement": 'P', "justification": 'Premise', "evidence": ''},
    {"stepID": 1, "statement": 'P→Q', "justification": 'Premise', "evidence": ''},
    {"stepID": 2, "statement": 'Q', "justification": 'Modus Ponens', "evidence": '1,2'}
]

print("Original Premises")
print(premises)
pretty_print_solution(solution)

premises = use_modus_tollens(premises, premises[0], solution)
print("\nModus Tollens Premises")
print(premises)
pretty_print_solution(solution)

premises = use_modus_ponens(premises, premises[0])
print("\nModus Ponens Premises")
print(premises)
pretty_print_solution(solution)

premises = use_disjunctive_syllogism(premises, premises[0])
print("\nDisjunctive Syllogism Premises")
print(premises)
pretty_print_solution(solution)

File: backend/compactGeneration.js


//give modusTollens, the desired conclusion = premise to replace and var2

function findUniqueAlphabetLetters(premises) {
    // Create a Set to store unique letters
    const uniqueLetters = new Set();

    // Iterate through the premises
    for (const premise of premises) {
        // Use a regular expression to match alphabet letters
        const letters = premise.match(/[A-Za-z]+/g);

        if (letters) {
            // Add the matched letters to the uniqueLetters Set
            for (const letter of letters) {
                uniqueLetters.add(letter);
            }
        }
    }

    return uniqueLetters;
}

function findXNonUniqueLetters(premises, x) {
    // Find the unique letters
    const uniqueLetters = findUniqueAlphabetLetters(premises);
    
    // Generate an array of non-unique capital letters
    const nonUniqueLetters = [];
    for (let letter = 'A'; nonUniqueLetters.length < x; letter = String.fromCharCode(letter.charCodeAt(0) + 1)) {
        if (!uniqueLetters.has(letter)) {
            nonUniqueLetters.push(letter);
        }
    }

    return nonUniqueLetters;
}


function toggleNegation(expression) {
    // Check if the expression starts with "¬" (negation)
    if (expression.startsWith("¬")) {
        // If it does, remove the "¬" symbol and return the variable without negation
        return expression.slice(1);
    } else {
        // If there's no "¬" symbol, add "¬" to negate the variable
        return expression.length == 1 ? "¬" + expression:"¬(" + expression + ")";
    }
}

function modusTollens(premise_to_replace, var2) {

    //if premise_to_replace = conclusion
    //var1 = not premise_to_replace
    // toggle ¬
    //var2 is unused letter

    //say premise_to_replace = P
    /////conclusion = premise_to_replace
    /////p = ¬var1
    /////var1 = (¬p)

    //say premsie_to_replace = p --> q
    /////conclusion = premise_to_replace
    /////p --> q = ¬var1
    /////var1 = ¬(p-->q)

    //The letters used in the premise_to_replace are known, var2 will be new letter
    //however code does NOT cancel double negatives

    let var1 = toggleNegation(premise_to_replace)
    return {
        premise1: toggleNegation(var2),
        premise2: `${var1}→${var2}`,
        conclusion: `${premise_to_replace}`
    }
}


function modusPonens(premise_to_replace, var2) {
    return {
        premise1: var2,
        premise2: `${var2}→(${premise_to_replace})`,
        conclusion: premise_to_replace
    }
}

function disjunctiveSyllogism(premise_to_replace, var2) {
    return {
        premise1: `${var2}∨(${premise_to_replace})`,
        premise2: toggleNegation(var2),
        conclusion: premise_to_replace
    }
}


function useModusTollens(premises, premise_to_replace, solution) {
    //find set of letters already in use
    let x = 1
    const nonUniqueLetters = findXNonUniqueLetters(premises, x);
    
    let var2 = nonUniqueLetters[0]

    let filledRule = modusTollens(premise_to_replace, var2)

    const indexToReplace = solution.findIndex((step) => step.statement === premise_to_replace);
    solution.splice(indexToReplace, 1, 
        {stepID: indexToReplace, statement:filledRule.premise1, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 1, statement: filledRule.premise2, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 2, statement: premise_to_replace, justification: 'Modus Tollens', evidence: `${indexToReplace + 1}, ${indexToReplace + 2}`},
    )

    for (let i = indexToReplace + 3; i < solution.length; i++) {
        solution[i].stepID = i
        if (solution[i].evidence !== '') {
            const numbers = solution[i].evidence.split(',').map(Number);
            // Add the value to each number
            const updatedNumbers = numbers.map((number) => {
                if (number >= indexToReplace + 1) {
                    return number + 2
                } else {
                    return number
                }
                //return number + 2 //when both of the lines referring to are shifted forward
            });
            // Join the updated numbers back into a string
            const updatedEvidenceString = updatedNumbers.join(',');
            solution[i].evidence = updatedEvidenceString
        }
    }

    // change the stepID of all proceeding steps
    
    let newPremises = premises.filter(premise => premise != premise_to_replace)
    newPremises.push(filledRule.premise1)
    newPremises.push(filledRule.premise2)

    return newPremises
}


function useModusPonens(premises, premise_to_replace) {
    //find set of letters already in use
    let x = 1
    const nonUniqueLetters = findXNonUniqueLetters(premises, x);
    let var2 = nonUniqueLetters[0]

    let filledRule = modusPonens(premise_to_replace, var2)

    const indexToReplace = solution.findIndex((step) => step.statement === premise_to_replace);
    solution.splice(indexToReplace, 1, 
        {stepID: indexToReplace, statement:filledRule.premise1, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 1, statement: filledRule.premise2, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 2, statement: premise_to_replace, justification: 'Modus Ponens', evidence: `${indexToReplace + 1}, ${indexToReplace + 2}`},
    )

    for (let i = indexToReplace + 3; i < solution.length; i++) {
        solution[i].stepID = i
        if (solution[i].evidence !== '') {
            const numbers = solution[i].evidence.split(',').map(Number);
            // Add the value to each number
            const updatedNumbers = numbers.map((number) => {
                if (number >= indexToReplace + 1) {
                    return number + 2
                } else {
                    return number
                }
                //return number + 2 //when both of the lines referring to are shifted forward
            });
            // Join the updated numbers back into a string
            const updatedEvidenceString = updatedNumbers.join(',');
            solution[i].evidence = updatedEvidenceString
        }
    }

    let newPremises = premises.filter(premise => premise != premise_to_replace)
    newPremises.push(filledRule.premise1)
    newPremises.push(filledRule.premise2)

    return newPremises
}


function useDisjunctiveSyllogism(premises, premise_to_replace) {
    //find set of letters already in use
    let x = 1
    const nonUniqueLetters = findXNonUniqueLetters(premises, x);
    let var2 = nonUniqueLetters[0]

    let filledRule = disjunctiveSyllogism(premise_to_replace, var2)

    const indexToReplace = solution.findIndex((step) => step.statement === premise_to_replace);
    solution.splice(indexToReplace, 1, 
        {stepID: indexToReplace, statement:filledRule.premise1, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 1, statement: filledRule.premise2, justification: 'Premise', evidence: ''},
        {stepID: indexToReplace + 2, statement: premise_to_replace, justification: 'Disjunctive Syllogism', evidence: `${indexToReplace + 1}, ${indexToReplace + 2}`},
    )

    for (let i = indexToReplace + 3; i < solution.length; i++) {
        solution[i].stepID = i
        if (solution[i].evidence !== '') {
            const numbers = solution[i].evidence.split(',').map(Number);
            // Add the value to each number
            const updatedNumbers = numbers.map((number) => {
                if (number >= indexToReplace + 1) {
                    return number + 2
                } else {
                    return number
                }
                //return number + 2 //when both of the lines referring to are shifted forward
            });
            // Join the updated numbers back into a string
            const updatedEvidenceString = updatedNumbers.join(',');
            solution[i].evidence = updatedEvidenceString
        }
    }
    let newPremises = premises.filter(premise => premise != premise_to_replace)
    newPremises.push(filledRule.premise1)
    newPremises.push(filledRule.premise2)

    return newPremises
}

let premises = ["P","P→Q",]
let conclusion = "Q"
let solution = [
    {stepID: 0, statement: 'P', justification: 'Premise', evidence: ''},
    {stepID: 1, statement: 'P→Q', justification: 'Premise', evidence: ''},
    {stepID: 2, statement: 'Q', justification: 'Modus Ponens', evidence: '1,2'}
]

console.log("Original Premises")
console.log(premises)
console.log(solution)

console.log("============================================")
premises = useModusTollens(premises, premises[0], solution)
console.log("Modus Tollens Premises")
console.log(premises)
console.log(solution)

console.log("============================================")
premises = useModusPonens(premises, premises[0])
console.log("Modus Ponens Premises")
console.log(premises)
console.log(solution)


console.log("============================================")
premises = useDisjunctiveSyllogism(premises, premises[0])
console.log("Disjunctive Syllogism Premises")
console.log(premises)
console.log(solution)
